### PR TITLE
Add uuidv4_text function

### PR DESCRIPTION
**Pull request by xsellier**
A the function `uuidv4_text()` to GDScript, and to the `String` class

It's been tested on Windows, Mac and Linux

<a href="https://github.com/godotengine/godot-proposals/issues/1654">Godot Proposal</a>

### CODE REVIEW

**Comment by Calinou** [comment]
Pull requests should be opened against the `master` branch then cherry-picked to the `3.2` branch.

**Comment by xsellier** [comment]
Sorry about

> Pull requests should be opened against the `master` branch then cherry-picked to the `3.2` branch, unless they don't apply to tHe `master` branch in the first place.

Sorry about that. I should have opened the PR for the `master` branch, however I don't have time to do so, so I opened it there, so maybe someone would pick it up, and port it to the master branch, then close this one and cherry-pick to the 3.2 branch.

File: modules/gdscript/doc_classes/@GDScript.xml [hunk 1]
```diff
@@ -1256,6 +1256,13 @@
 				[/codeblock]
 			</description>
 		</method>
+		<method name="uuidv4_text">
+			<return type="String">
+			</return>
+			<description>
+				Returns a uuidv4 string
```
Comment (Calinou) [inline modules/gdscript/doc_classes/@GDScript.xml:1263]:
```suggestion
				Returns a random 128-bit [url=https://en.wikipedia.org/wiki/Universally_unique_identifier]universally unique identifier[/url]. Example returned value: [code]dd10eb87-1596-426f-9e1d-496e8ceea6ed[/code].
```

**Review by Calinou — CHANGES_REQUESTED** [review]

**Comment by Xrayez** [comment]
> > Pull requests should be opened against the `master` branch then cherry-picked to the `3.2` branch, unless they don't apply to tHe `master` branch in the first place.
> 
> Sorry about that. I should have opened the PR for the `master` branch, however I don't have time to do so, so I opened it there, so maybe someone would pick it up, and port it to the master branch, then close this one and cherry-pick to the 3.2 branch.

For reference, this is what I've been talking about in: https://github.com/godotengine/godot/pull/39871#issuecomment-650533554.

**Comment by vnen** [comment]
Is there any good reason to add this? It could be done with a script.

**Comment by xsellier** [comment]
I'm using for:
- Custom database
- Analytics
- Bug report

Why it is done via C++ and not GDScript, because the C++ function outperforms GDScript (10x faster).

**Comment by vnen** [comment]
Sure it's faster, but it doesn't look like performance critical code and you could do it with GDNative to get better performance.

**Comment by Zireael07** [comment]
If he's using it for a database, then it's absolutely performance critical.

**Comment by Calinou** [comment]
@Zireael07 In a database, you usually have to perform this UUIDv4 conversion only every time you insert an element. Most databases don't even get close to 100 inserts a second :slightly_smiling_face:

**Comment by xsellier** [comment]
Getting a unique identifier accross the world is useful for a game. Specially if you make a multiplayer game.
About performances matters. Indeed everybody can implement the uuidv4 function on their side, or it could be shared via an addon. It is more convenient to have the uuidv4 function builtin, than having to maintain an addon accross all the godot version.

**Comment by Calinou** [comment]
> than having to maintain an addon accross all the godot version (that are not backward compatible)

That's what Git branches are for :slightly_smiling_face:

**Comment by xsellier** [comment]
> That's what Git branches are for slightly_smiling_face
> 
> Such an add-on doesn't require a great amount of maintenance. It doesn't interact with any of the editor.

Why add that work when you can make it built-in and would require no work at all?

**Comment by Calinou** [comment]
> Why add that work when you can make it built-in and would require no work at all?

When we add something to core, it shifts maintenance from the user to core developers. However, we only have a finite amount of manpower.

**Comment by xsellier** [comment]
> When we add something to core, it shifts the maintenance burden from the user to core developers. However, we only have a finite amount of manpower.

Regarding `uuidv4_text`, it relies on 3 other functions: `Math::rand`, `OS::get_unix_time` and `OS::get_ticks_msec`.
According to git, [those](https://github.com/godotengine/godot/blame/master/core/math/math_funcs.cpp#L54) [functions](https://github.com/godotengine/godot/blame/master/core/os/os.cpp#L50) [haven't](https://github.com/godotengine/godot/blame/master/core/os/os.cpp#L83) changed a lot. So it wouldn't add much work (or even no work at all) to the core developpers.

**Comment by Xrayez** [comment]
> Once we add it, we essentially commit to supporting it forever.

Nothing is eternal. #42113 is a clear example of this.

> However, we only have a finite amount of manpower.

From my observations, we keep gaining more contributors, not less, I'd safely extrapolate that at least 1% of those contributors will become one of those developers who can maintain the engine.

> That's what Git branches are for 🙂

There are generally many other smaller features like this which are difficult to keep feature branches for. One has to either use a patching system and/or maintain a custom version of the engine to have persistent, reproducible core functionality across game projects (which *also* require performance), but that's only applicable for in-house solutions...

Also, I've seen a tendency with Godot development that even the smallest features attract controversial discussions. I don't understand this. I believe this PR is general-purpose enough to cater most needs. By "most" I mean is that, at some point of development, features like this are ***essential to have in any big project with enough of playerbase***, which may not be useful for beginner developers *just yet*.

Talking about the "adoption" by game development companies across the world as desired by some core developers, alas the current development approach and evaluation of features don't seem to facilitate this.

If Godot Engine caters mostly to beginner use cases, then this should be written on the [Godot Engine homepage](https://godotengine.org/). But do note that the current wording doesn't reflect this:

> Godot provides a ***huge*** **set of common tools** [emphasis mine], so you can just focus on making your game without reinventing the wheel.

**Comment by Yukitty** [comment]
This should use crypto rand instead of math rand, and the UUID v4 standard does not include timestamps (those are used by other UUID versions)

**Comment by PoqXert** [comment]
Why you add function to `String` class? Is UUID based on string? What if you want to get this in bytes?

I think that you need to create `UUID` class with functions for creation UUID v4 and getting it as string. (Other UUID versions may be added to this class.)
For example:
```gdscript
var id := UUID.create_v4();
var id_text := id.to_string();
```
You can create it as a module and request addition to default modules.

**Comment by xsellier** [comment]
> This should use crypto rand instead of math rand, and the UUID v4 standard does not include timestamps (those are used by other UUID versions)

RFC for UUID v4:
https://www.ietf.org/rfc/rfc4122.txt

>      UUID                   = time-low "-" time-mid "-"
>                               time-high-and-version "-"
>                               clock-seq-and-reserved
>                               clock-seq-low "-" node
>      time-low               = 4hexOctet
>      time-mid               = 2hexOctet
>      time-high-and-version  = 2hexOctet
>      clock-seq-and-reserved = hexOctet
>      clock-seq-low          = hexOctet
>      node                   = 6hexOctet
>      hexOctet               = hexDigit hexDigit
>      hexDigit =
>            "0" / "1" / "2" / "3" / "4" / "5" / "6" / "7" / "8" / "9" /
>            "a" / "b" / "c" / "d" / "e" / "f" /
>            "A" / "B" / "C" / "D" / "E" / "F"

Later in this doc

> 4.1.2.  Layout and Byte Order
> 
>    To minimize confusion about bit assignments within octets, the UUID
>    record definition is defined only in terms of fields that are
>    integral numbers of octets.  The fields are presented with the most
>    significant one first.
> 
>    Field                  Data Type     Octet  Note
>                                         #
> 
>    time_low               unsigned 32   0-3    The low field of the
>                           bit integer          timestamp
> 
>    time_mid               unsigned 16   4-5    The middle field of the
>                           bit integer          timestamp
> 
>    time_hi_and_version    unsigned 16   6-7    The high field of the
>                           bit integer          timestamp multiplexed
>                                                with the version number

**Comment by xsellier** [comment]
> Why you add function to `String` class? Is UUID based on string? What if you want to get this in bytes?
> 
> I think that you need to create `UUID` class with functions for creation UUID v4 and getting it as string. (Other UUID versions may be added to this class.)
> For example:
> 
> ```gdscript
> var id := UUID.create_v4();
> var id_text := id.to_string();
> ```
> 
> You can create it as a module and request addition to default modules.

Since the String class has hash, md5 and sha functions I though it would be a good candidate for uuid:
https://docs.godotengine.org/fr/latest/classes/class_string.html#class-string-method-hash

**Comment by xsellier** [comment]
Since this PR is dead and it is not going to be merged, Im closing it. I cnnot afford spending time answering all the question you are asking.

**Pull request closed without merging**